### ISSUE TITLE
Compilation often fails on gocryptfs

### ISSUE THREAD

**Issue by krumelmonster** [rust-lang/rust#93900]
I often get these errors when trying to compile any rust project on a gocryptfs filesystem (gocryptfs v2.2.1; go-fuse [vendored]; 2021-10-30 go1.17.2 linux/amd64). There have been such problems with nightly, beta and stable.

```
$ cargo build
   Compiling hello v0.1.0 (/home/c/rust-hello)
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `82`,
 right: `1002111927320821928687967599834759150`', compiler/rustc_query_impl/src/on_disk_cache.rs:547:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.59.0-beta.6 (0426998f5 2022-02-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `hello`
```

and

```
$ cargo build
   Compiling hello v0.1.0 (/home/c/rust-hello)
error: could not compile `hello`

Caused by:
  process didn't exit successfully: `rustc --crate-name hello --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi,future-incompat --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 -C metadata=2c3affb4eb36b7d2 -C extra-filename=-2c3affb4eb36b7d2 --out-dir /home/c/rust-hello/target/debug/deps -C incremental=/home/c/rust-hello/target/debug/incremental -L dependency=/home/c/rust-hello/target/debug/deps` (signal: 7, SIGBUS: access to undefined memory)
```

**Comment by csmoe** [comment]
@krumelmonster would you mind sharing your code here?

**Comment by krumelmonster** [comment]
Any code I tested, for example https://github.com/spiegel-im-spiegel/rust-hello

**Comment by saethlin** [comment]
As far as I can tell, the ICE reported in this issue is caused by another process damaging rustc's build artifacts or by a buggy filesystem. That other process could be anything from an IDE running the compiler at the same time as a manual build to a find-and-replace that accidentally edits files in the target directory.

https://github.com/rust-lang/rust/pull/124686 contains a mitigation for this situation; the compiler now has a crude way to detect and ignore damaged incremental compilation files. You will still get a warning about the damaged file, but the build should proceed normally. That change shipped in Rust 1.80 which released today, so I am closing this issue and a few others which as far as I can tell all have the same root cause. You shouldn't see this ICE anymore, but please file a new issue if you see something like it on 1.80 or later.
